source/util/parse_number.cpp

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "util/parse_number.h"
+
+#include <functional>
+#include <iomanip>
+#include <sstream>
+#include <string>
+#include <tuple>
+
+#include "util/hex_float.h"
+
+namespace spvutils {
+std::tuple<EncodeNumberStatus, std::string> ParseAndEncodeIntegerNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit) {
+  std::ostringstream error_msg_stream;
```

Comment:
This object is created on the success path. Before the diagnostic object was only created on the error paths. 
I can live with this for now.  At least it's not forcing wastefulness at the API boundary.  Later we can rework this to only create the object on the error path, by making something like the DiagnosticStream here (Converts to the status type, but can accumulate a string and emit it to a particular destination.)